irohad/main/impl/ordering_init.cpp

```diff
@@ -6,18 +6,31 @@
 #include "main/impl/ordering_init.hpp"
 #include "ametsuchi/ordering_service_persistent_state.hpp"
 #include "interfaces/common_objects/peer.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "interfaces/iroha_internal/block.hpp"
 
 namespace iroha {
   namespace network {
     auto OrderingInit::createGate(
         std::shared_ptr<OrderingGateTransport> transport,
         std::shared_ptr<ametsuchi::BlockQuery> block_query) {
-      auto height = block_query->getTopBlocks(1).as_blocking().last()->height();
-      auto gate =
-          std::make_shared<ordering::OrderingGateImpl>(transport, height);
-      log_->info("Creating Ordering Gate with initial height {}", height);
-      transport->subscribe(gate);
-      return gate;
+      return block_query->getTopBlock().match(
+          [this, &transport](
+              expected::Value<std::shared_ptr<shared_model::interface::Block>>
+                  &block) -> std::shared_ptr<ordering::OrderingGateImpl> {
```

Comment:
Please change return type of the lambda to std::shared_ptr<ordering::OrderingGate> to preserve polymorphism